### PR TITLE
refs #25: ability to set the widget style class through the view DSL

### PR DESCRIPTION
**Pull request by emmanueltouzery**
I've wanted this feature in relm for a while, in my relm apps i have a bunch of `get_style_context().add_class()` calls in my `init_view` implementations.

The new `style_class` attribute must enable to specify multiple style classes for a widget.. currently I made it that you must repeat the attribute multiple times in the DSL. We could also change it so that the user can type `#[style_class=("destructive-action", "linked")]`. Let me know if you'd prefer that.

I also had to handle style classes specially in the `Attributes` struct, because with a Map of attributes, we couldn't list multiple times the style_class attribute. Again other approaches are possible, if you'd rather I do it another way, let me know!

### CODE REVIEW

**Comment by antoyo** [comment]
Nice. Have you got an idea for specifying the CSS file?

Maybe we could have an additional attribute like `#[css_file="style.css"]` that would expand to:
```rust
        use gtk::{CssProviderExt, StyleContextExt};
        let style_context = self.button.get_style_context();
        let style = include_bytes!("style.css");
        let provider = gtk::CssProvider::new();
        provider.load_from_data(style).unwrap();
        style_context.add_provider(&provider, gtk::STYLE_PROVIDER_PRIORITY_APPLICATION);
```
What do you think?

**Comment by emmanueltouzery** [comment]
hmm, for loading the style, that's not exactly the code that I use. Your code is tied to a widget, mine is tied to a screen:

```rust
        let screen = self.widgets.window.get_screen().unwrap();
        let css = gtk::CssProvider::new();
        css.load_from_data(CSS_DATA)?;
        gtk::StyleContext::add_provider_for_screen(
            &screen,
            &css,
            gtk::STYLE_PROVIDER_PRIORITY_APPLICATION,
        );
```

Yours is probably better. But in any case we need a widget to work with, you just use it directly and i use it for the screen (in theory i guess it could be bad for multiscreen, but it works for me..).

In general, why not. I wasn't bothering too much with this, because you need this only once in the application. In terms of implementation, so that we don't "inject" too much code through that macro, maybe `relm` could export a function `bind_css_file(widget: &gtk::Widget, css_contents: &str)` or something, and then the macro would just generate something like `bind_css_file(widget, include_bytes!(css_path)`. But we could also generate the whole code if you prefer.

If you specify a way to do it, I could implement it.

**Comment by antoyo** [comment]
That's a good point: let's just merge as is and we'll think about what could be the best way to do this.

File: relm-derive/src/gen/parser.rs [hunk 1]
```diff
@@ -400,25 +404,35 @@ impl Parse for Attribute {
...
         loop {
             let lookahead = input.lookahead1();
 
             if lookahead.peek(Token![#]) {
                 let attribute: Attribute = input.parse()?;
-                name_values.extend(attribute.name_values);
+                match attribute.name_values.get("style_class") {
+                    Some(Some(style_class)) => {
+                        style_classes.insert(style_class.value());
+                        ()
```
Comment (antoyo) [inline relm-derive/src/gen/parser.rs:422]: This line should not be necessary.

File: relm-derive/src/gen/gen.rs [hunk 2]
```diff
@@ -388,6 +388,9 @@ impl<'a> Generator<'a> {
         let ident = quote! { #widget_name };
         let (properties, visible_properties) = self.gtk_set_prop_calls(widget, ident);
         let child_properties = gen_set_child_prop_calls(widget, parent, parent_widget_type, IsGtk);
+        let set_style_classes: Vec<_> = widget.style_classes.iter().map(|style_class|
+            quote_spanned! { widget_name.span() => gtk::StyleContextExt::add_class(&#widget_name.get_style_context(), &#style_class); }
```
Comment (antoyo) [inline relm-derive/src/gen/gen.rs:392]: Is this the right span?
What happens when the trait `WidgetExt` is not in scope?

Also, could you check that the error is at the right location when you provide something else than a string in the `style_class` attribute?

**Review by antoyo — CHANGES_REQUESTED** [review]

File: relm-derive/src/gen/parser.rs (hunk 1 above)
Comment (emmanueltouzery) [reply relm-derive/src/gen/parser.rs:422]: you mean the `()`? the problem is that the `match` wants every branch to return the same type, and the other branch returns `()` while the insert returns `bool`. I quite believe there's a nicer way to fix that, but I don't know it.

Comment (antoyo) [reply relm-derive/src/gen/parser.rs:422]: Yeah, I've seen this issue, but usually, having the curly braces and the semi-colon should be enough. Are you sure it's not working when you remove the `()`?

Comment (emmanueltouzery) [reply relm-derive/src/gen/parser.rs]: you are right, i could just remove it, pushed now in a followup commit.

File: relm-derive/src/gen/gen.rs (hunk 2 above)
Comment (emmanueltouzery) [reply relm-derive/src/gen/gen.rs:392]: > is this the right span?

I think so, all the code around uses that span, for instance to generate the `widget.show()` and the entire block. In the end this `set_style_classes` is embedded in another quote line 405, which also uses that same span.

> What happens when the trait `WidgetExt` is not in scope?

I think it's OK. I call [add_class(&self, class_name: &str)](https://gtk-rs.org/docs/gtk/trait.StyleContextExt.html#tymethod.add_class) "manually" on `gtk::StyleContextExt`, so we don't need StyleContextExt to be in scope. In addition, I added this attribute in the `buttons-attribute.rs` test, and there the imports really seem minimal to me, and it works fine.

> Also, could you check that the error is at the right location when you provide something else than a string in the style_class attribute?

I tried it in the in the `buttons-attribute.rs` example, and this is the compilation error if I change:

```diff
 -               #[style_class="linked"]
 +               #[style_class=linked]
```

```
error: expected `,`
  --> examples/buttons-attribute.rs:88:31
   |
88 |                 #[style_class=linked]
   |                               ^^^^^^

error[E0433]: failed to resolve: use of undeclared type `Win`
   --> examples/buttons-attribute.rs:100:5
    |
100 |     Win::run(()).expect("Win::run failed");
    |     ^^^ use of undeclared type `Win`

warning: unused imports: `ButtonExt`, `Inhibit`, `LabelExt`, `OrientableExt`, `WidgetExt`
  --> examples/buttons-attribute.rs:23:5
   |
23 |     ButtonExt,
   |     ^^^^^^^^^
24 |     Inhibit,
   |     ^^^^^^^
25 |     LabelExt,
   |     ^^^^^^^^
26 |     OrientableExt,
   |     ^^^^^^^^^^^^^
27 |     WidgetExt,
   |     ^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused import: `gtk::Orientation::Vertical`
  --> examples/buttons-attribute.rs:29:5
   |
29 | use gtk::Orientation::Vertical;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused import: `relm::Widget`
  --> examples/buttons-attribute.rs:30:5
   |
30 | use relm::Widget;
   |     ^^^^^^^^^^^^

warning: unused import: `self::Msg::*`
  --> examples/buttons-attribute.rs:33:5
   |
33 | use self::Msg::*;
   |     ^^^^^^^^^^^^

error: aborting due to 2 previous errors; 4 warnings emitted

For more information about this error, try `rustc --explain E0433`.
error: could not compile `relm-examples`

To learn more, run the command again with --verbose.
```

This look good to me. However I also tried this:

```
                #[style_class=12]
```

and in that case the error message is not great:

```
error: custom attribute panicked
  --> examples/buttons-attribute.rs:49:1
   |
49 | #[widget]
   | ^^^^^^^^^
   |
   = help: message: Invalid style_class specification
```

However I'm not really sure what I can do to improve this :(
This is now unrelated to this code, it's related to the parsing code, and I think to parts of it that I didn't touch.

**Comment by antoyo** [comment]
Sounds good!
Is it ready to merge?

**Comment by emmanueltouzery** [comment]
I am happy with it yes, from my point of view it's ready to merge, ONLY i'd be ready to add some documentation somewhere, but I'm not sure where. As you saw I updated an example, that may be enough.

Also I highlighted some choices to make in my initial message, this choice is related to the API, so you might consider whether you're happy with it as-is:

> The new style_class attribute must enable to specify multiple style classes for a widget.. currently I made it that you must repeat the attribute multiple times in the DSL. We could also change it so that the user can type #[style_class=("destructive-action", "linked")]. Let me know if you'd prefer that.

**Review by antoyo — APPROVED** [review]

**Comment by antoyo** [comment]
That's fine: we can always change it later if needed.

For the documentation, that would eventually go in [this page](https://relm.antoyo.xyz/documentation/reference/). If you're willing to try, you could start adding content to it (https://github.com/antoyo/jekyll-relm-blog/blob/master/documentation/reference.adoc), perhaps by doing something similar to the [tutorial page](https://github.com/antoyo/jekyll-relm-blog/blob/master/documentation/tutorial.adoc).
I can help you if needed.

**Comment by emmanueltouzery** [comment]
ah, short-term i had in mind to expand some documentation to mention this new feature, writing or starting to write the reference from scratch is considerably more work. i do really really like relm though.. maybe at some point i try... but not short-term i think.

**Comment by antoyo** [comment]
If it would help you that I start the reference so you can just add the documentation for this feature, please tell me.

**Comment by emmanueltouzery** [comment]
I'm not sure how the reference should look like. There is the tutorial and the crates.io apidoc.

Maybe some sort of description of the `#[widget]`  macro, `view!`, any supported annotations? So, `name`, `style_class`, maybe some other thing?

**Comment by antoyo** [comment]
Yep, that's the plan.

**Comment by emmanueltouzery** [comment]
i'm working on something, i'll make a PR on that other repo I think tonight.

**Pull request merged**